---
title: "cTWAS vs FOCUS"
author: "Julian McClellan"
date: "2023-08-28"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

```{r setup, message=F, include=F}
library(tidyverse)
library(glue)
library(cowplot)
library(purrr)
knitr::opts_chunk$set(echo = FALSE)
# https://stackoverflow.com/questions/30237310/setting-work-directory-in-knitr-using-opts-chunksetroot-dir-doesnt-wor
knitr::opts_knit$set(root.dir = file.path(rprojroot::find_rstudio_root_file()))
options(dplyr.summarise.inform = FALSE)

# https://github.com/rstudio/DT/issues/67
# You need this code to conduct the magic dependences attaching...
DT::datatable(matrix())

source("code/prep_data.R")
source("code/locus_plot.R")
WANT_STUDIES <- c("meta_analysis_BCAC_UKB_ovr_brca")
prepped_data <- list()
for (study in WANT_STUDIES){
  prepped_data[[study]] <- runonce::save_run({main(study)}, glue("output/cache/{study}.rds"))
  attach(prepped_data[[study]])
}
  
#   rlist <- list(top_mismatch_genes = top_mismatch_genes,
#                 mismatch_gene_tissue_tracker = mismatch_gene_tissue_tracker,
#                 want_tissues=want_tissues,
#                 ctwas_res=ctwas_res,
#                 focus_low_cutoff=focus_low_cutoff,
#                 ctwas_high_cutoff=ctwas_high_cutoff
#                 )
wt_idx <- 1


do_tissue <- function(wt_idx){
  # First a table showing all mismatching genes
  do_chrom_region <- function(chrom, region_tag2){
    # First the locus plot
    cur_tiss_chrom_region_mismatch_tracker <- cur_tiss_chrom_mismatch_tracker %>%
      filter(region_tag2 == !!region_tag2)

    ctwas_res_filt <- ctwas_res[[want_tissues[wt_idx]]] %>%
      filter(chrom == !!chrom, region_tag2 == !!region_tag2)
    region_tag_pos_min <- ctwas_res_filt %>% pull(pos) %>% min(., na.rm=T)
    region_tag_pos_max <- ctwas_res_filt %>% pull(pos) %>% max(., na.rm=T)
    
    pre_fig_table <- DT::datatable(cur_tiss_chrom_region_mismatch_tracker %>% 
                  select(all_of(c("gene_id.v26", "gene_name.v40", "tiss_gwide_twas_bfr_sig", "tiss_bfr_thresh"))),
              caption=glue("Chr{chrom} | ~ {region_tag_pos_min}-{region_tag_pos_max}"))

    cat("\n\n")
    cat(knitr::knit_print(pre_fig_table))
    cat("\n\n")
  
    label_genes <- cur_tiss_chrom_region_mismatch_tracker %>%
      pull(`gene_id.v26`)
  
    a <- locus_plot(WANT_STUDIES[1],
                    want_tissues[wt_idx],
                    ctwas_res_filt,
                    chrom=chrom,
                    region_tag2=region_tag2,
                    eqtl_or_sqtl="eqtl",
                    gwide_tiss_sig_thresh=gwide_tiss_sig_thresh,
                    return_table=F,
                    focus=NULL,
                    label_genes=label_genes,
                    rerun_ctwas=F,
                    rerun_load_only=F,
                    label_panel="both",
                    legend_side="left",
                    legend_panel="",
                    draw_gene_track = F)
  }
  cur_tiss_mismatch_tracker <- mismatch_gene_tissue_tracker %>%
    filter(tissue == want_tissues[wt_idx])
  
  gwide_tiss_sig_thresh <- cur_tiss_mismatch_tracker %>%
    filter(tissue == want_tissues[wt_idx]) %>%
    pull(tiss_bfr_thresh) %>%
    max(., na.rm=T)
  
  cat("\n\n")
  cat(knitr::knit_print(DT::datatable(cur_tiss_mismatch_tracker %>%
                  arrange(chrom, region_tag2) %>%
                  select(all_of(c("gene_id.v26", "gene_name.v40", "chrom", "tiss_gwide_twas_bfr_sig"))),
                            caption=glue("{want_tissues[wt_idx]}: FOCUS PIP < {focus_low_cutoff} & cTWAS PIP > {ctwas_high_cutoff}")
              )))
  cat("\n\n")

  cur_tiss_chrom_region_df <- cur_tiss_mismatch_tracker %>%
    group_by(chrom, region_tag2) %>%
    summarize() %>%
    head(2) # debug

  for (cur_chr in cur_tiss_chrom_region_df %>% pull(chrom) %>% unique() %>% sort()){
    cur_diag_graph <- readRDS(glue("output/cache/ctwas_diag_{WANT_STUDIES[1]}__{want_tissues[wt_idx]}__eqtl_chr{cur_chr}.rds"))
    print(glue("Chromosome {cur_chr}. {want_tissues[wt_idx]}"))
    print(cur_diag_graph)
  
    cur_tiss_chrom_mismatch_tracker <- cur_tiss_mismatch_tracker %>%
      filter(chrom == !!cur_chr)
  
    cur_tiss_chrom_region_df %>%
      filter(chrom == !!cur_chr) %>%
      pwalk(do_chrom_region)
  }
}
```

## Introduction

```{r introduction, message=F, warning=F}
DT::datatable(top_mismatch_genes)
```

## By Tissue Figures and Tables

```{r iterate_tissues, message=F,warning=F, include=F}
# https://stackoverflow.com/questions/21729415/generate-dynamic-r-markdown-blocks
out = NULL
for (i in 10:11){
  out <- c(out, knitr::knit_expand("analysis/chunks/tissue_chunk.Rmd"))
}
```

`r paste(knitr::knit(text=out), collapse='\n\n')`

### `r want_tissues[wt_idx]`

Genome wide Bonferroni threshold for `r want_tissues[wt_idx]`: `r mismatch_gene_tissue_tracker %>% filter(tissue == want_tissues[wt_idx]) %>% pull(tiss_bfr_thresh) %>% max(., na.rm=T)`.

```{r tiss, results='asis', message=T,warning=F}
do_tissue(wt_idx)
wt_idx <- wt_idx + 1
```

 #### `r want_tissues[wt_idx]` ### `r want_tissues[3]`
 #### `r want_tissues[wt_idx]` ### `r want_tissues[4]`
 #### `r want_tissues[wt_idx]` ### `r want_tissues[5]`
 #### `r want_tissues[wt_idx]` ### `r want_tissues[6]`
 #### `r want_tissues[wt_idx]` ### `r want_tissues[7]`
 #### `r want_tissues[wt_idx]` ### `r want_tissues[8]`
 #### `r want_tissues[wt_idx]` ### `r want_tissues[9]`
 #### `r want_tissues[wt_idx]` ### `r want_tissues[10]`
 #### `r want_tissues[wt_idx]` ### `r want_tissues[11]`

```{r test, include=F, echo=T}
# for (chr in 1:22){
#   cur_diag_graph <- readRDS(glue("output/cache/ctwas_diag_meta_analysis_BCAC_UKB_ovr_brca__Breast_Mammary_Tissue__eqtl_chr{chr}.rds"))
#   print(cur_diag_graph)
# }
```

```{r test2, include=F}
# knitr::include_graphics("assets/ctwas_diag_meta_analysis_BCAC_UKB_ovr_brca__Breast_Mammary_Tissue__eqtl_chr1.png", error=F)
# knitr::include_graphics("assets/ctwas_diag_meta_analysis_BCAC_UKB_ovr_brca__Breast_Mammary_Tissue__eqtl_chr2.png", error=F)
```

```{r test_locus_plot, message=F, warning=F}
# prepped_data <- list()
# for (study in WANT_STUDIES){
#   prepped_data[[study]] <- runonce::save_run({main(study)}, glue("output/cache/{study}.rds"))
# }
# a <- locus_plot(WANT_STUDIES[1], "Breast_Mammary_Tissue", prepped_data[["meta_analysis_BCAC_UKB_ovr_brca"]][["ctwas_res"]],
#                 
#                       eqtl_or_sqtl="eqtl",
#                       return_table=T,
#                       focus=NULL,
#                       label_genes=NULL,
#                       rerun_ctwas=F,
#                       rerun_load_only=F,
#                       label_panel="both",
#                       legend_side="left",
#                       legend_panel="",
#                       draw_gene_track = F)
```

